### PR TITLE
ClipToGrid funtionality

### PR DESCRIPTION
**Pull request by lossyrob**
This PR adds `ClipToGrid`, which takes an RDD of Features and clips them to `SpatialKey`s based on `LayoutDefinition`.

This functionality is generally useful, but was moved from a version in [VectorPipe](https://github.com/geotrellis/vectorpipe) to here in service of a specific class of vector join problems, in service of #2116. I [added and example](https://github.com/locationtech/geotrellis/pull/2424/files#diff-d80a3e82172d43a81e629e86465decbf) in the `doc-examples` project that shows a polygonal summary of `PolygonFeature[UUID]`s that were in an RDD, based on a set of GeoTiffs on S3.

A full solution to #2116 will take some more thought, which I hear is in line with some work the GeoMesa folk are doing, and we can collaborate on.

Some other things to note:
- I broke out some of the logic in the LayerFilters for MultiPolygon and MultiLine so I could reuse that logic. It seemed like the most appropriate place for it was MapKeyTransform.
- I fixed up a deprecation warning message along the way.

### CODE REVIEW

File: spark/src/main/scala/geotrellis/spark/tiling/MapKeyTransform.scala [hunk 1]
```diff
@@ -122,4 +126,55 @@ class MapKeyTransform(val extent: Extent, val layoutCols: Int, val layoutRows: I
...
+     * array of  tuples which  is then  mapped-over to  produce an
+     * array of KeyBounds.
+     */
+    val tiles = new ConcurrentHashMap[(Int,Int), Unit]
+    val fn = new Callback {
+      def apply(col : Int, row : Int): Unit = {
+        val tile : (Int, Int) = (bounds.colMin + col, bounds.rowMin + row)
+        tiles.put(tile, Unit)
+      }
+    }
```
Comment (echeipesh) [inline spark/src/main/scala/geotrellis/spark/tiling/MapKeyTransform.scala]: Some dragons been found around these parts, better to use this method: https://github.com/locationtech/geotrellis/blob/5c5ebed09f66cde92c77d9f21ce22e65a44573c4/spark/src/main/scala/geotrellis/spark/tiling/LayoutDefinition.scala#L38

Comment (fosskers) [reply spark/src/main/scala/geotrellis/spark/tiling/MapKeyTransform.scala]: Whoa, that's a sexy way to quote the line.

Comment (lossyrob) [reply spark/src/main/scala/geotrellis/spark/tiling/MapKeyTransform.scala]: @echeipesh I'm using the logic from our LayerFilters for MultiPolygon and MultiLine (which the Line and Polygon filters also use). If this is wrong, then they are also (and have been) wrong. I thought we sorted out all the dragons there - are there some specifics about the dragons?

**Comment by lossyrob** [comment]
Need to investigate and benchmark a way to have an efficient and correct line rasterization implementation for spatial keys. That bumps this PR past 1.2. :opposite of tada:

**Pull request closed without merging**